bigquery/gcloud/aio/bigquery/utils.py

```diff
@@ -0,0 +1,119 @@
+import datetime
+import logging
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import List
+
+
+log = logging.getLogger(__name__)
+
+
+def flatten(x: Any) -> Any:
+    """
+    Flatten response objects into something we can actually work with.
+
+    The API returns data of the form:
+
+        {'f': [{'v': ...}]}
+
+    to indicate groupings of fields and their potential for having multiple
+    values. We want those to just be plain old objects.
+    """
+    if isinstance(x, dict):
+        # TODO: what if a user has stored a dictionary in their table and that
+        # dictionary is shaped the same way as Google's response format?
+        if 'f' in x:
+            return [flatten(y['v']) for y in x['f']]
+
+        if 'v' in x:
+            return flatten(x['v'])
+
+    if isinstance(x, list):
+        return [flatten(y) for y in x]
+
+    return x
+
+
+def parse(field: Dict[str, Any], value: Any) -> Any:
+    """
+    Parse a given field back to a Python object.
+
+    This is often trivial: convert the value from a string to the type
+    specified in the field's schema. There's a couple caveats we've identified
+    so far, though:
+
+    * NULLABLE fields should be handled specially, eg. so as not to
+      accidentally convert them to the schema type.
+    * REPEATED fields are nested a biot differently than expected, so we need
+      to flatten *first*, then convert.
+
+    `Field = Dict[str, Union[str, 'Field']]`, but wow is that difficult to
+    represent in a backwards-enough compatible fashion.
+    """
+    try:
+        f: Callable[[Any], Any] = {  # type: ignore[assignment]
+            'BOOLEAN': lambda x: x == 'true',
+            'BYTES': bytes,
+            'FLOAT': float,
+            'INTEGER': int,
+            'NUMERIC': float,
+            'RECORD': dict,
+            'STRING': str,
+            'TIMESTAMP': lambda x: datetime.datetime.fromtimestamp(float(x)),
+        }[field['type']]
+    except KeyError:
+        # TODO: determine the proper methods for converting the following:
+        # BIGNUMERIC
+        # DATE
+        # DATETIME
+        # GEOGRAPHY
+        # TIME
```

Comment:
Can we leave them in the format returned by the API itself?

Maybe converting to a loss less datatype might be sufficient? For instance, int, str, str (both date times in ISO 8601), [unsure] , str?

Reply:
bignumeric -> int wouldn't work IIUC as bignumeric can hold larger values; I don't want to silently change folks' data on them. Converting these all to str imight be a decent idea, though... I guess I'm torn given `datetime` is probably the "right" answer. Is it better to convert them to string now and introduce a breaking change later once we have the sample data for converting 'em, or to introduce new support for additional fields via point releases? I'm leaning to the later.

Reply:
Ah, sorry, my mistake -- I was thinking of precision rather than max value. NUMERIC and BIGNUMERIC types have [38 and ~77 decimal places of precision](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types) respectively, whereas python floats are 18 places.

I guess that means I should use [decimal](https://docs.python.org/3/library/decimal.html) to do this properly.